### ISSUE TITLE
Suboptimal segmentation for RoPE

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/Fuser#2599]
# TL;DR

the RoPE module in Llama-2 (and possibly other configs as well) is over-segmented into six pointwise kernels. Two should suffice. This issue blocks https://github.com/Lightning-AI/lightning-thunder/pull/731, although we don't necessarily have to reach two to unblock. 

# To reproduce

```
$ pwd
/opt/pytorch/lightning-thunder
$ git checkout wjy/bookend
$ NVFUSER_DUMP=segmented_fusion pytest thunder/benchmarks/targets.py -k test_litgpt_qkv_split_rope[Llama-2-13b-hf-forward-bs2-thunder] -s | grep 'g{(pointwise' | wc -l
6
```

# Problem

The following figure shows the current suboptimal segmentation

![RoPE](https://github.com/user-attachments/assets/d62bd143-e2aa-4232-a3ca-ce493cd131e4)

The highlighted ops form the three pointwise kernels for Q. There's a similar pattern for K that's omitted for simplicity, so there are in total six pointwise kernels. The remaining, meta ops are expectedly segmented out as "no-op" regions. 

Note: For now, I omitted the `Pad`s before a `Cat` for simplicity. Soonish, https://github.com/NVIDIA/Fuser/pull/2373, pending, may move `Pad`s upstream, affecting segmentation. 

There are two main reasons we reached this state. 
1. The segmenter enforces a segment boundary before `Slice` and `Pad` (and thus before `Cat`). 
2. The `Squeeze` and the red `ToFloat` were merged too early. This made it impossible to keep merging the red and the green segment, because `Squeeze` is a transitive input to the `Cat` and the merge would lead to a cycle. Several sub-reasons contributed to that order: 
   - The merging runs [in topological order](https://github.com/NVIDIA/Fuser/blob/83e6de805e42b3063adc20eefcd02cfd5e8c1959/csrc/fusion_segmenter.cpp#L3733-L3734). 
   - The segmenter merges one edge at a time. When attempting to merge into the green `+` its operands, the segmenter only merged the right operand instead of both. 
   - A newly merged segment is pushed to [the end](https://github.com/NVIDIA/Fuser/blob/83e6de805e42b3063adc20eefcd02cfd5e8c1959/csrc/fusion_segmenter.cpp#L745) of the worklist.

**Comment by wujingyue** [comment]
The attempted fix had a mixed effect on performance: https://github.com/NVIDIA/Fuser/pull/2815#issuecomment-2319068605. So this is still open. I'm reassigning this to @jjsjann123, the POC for RoPE.

**Comment by jjsjann123** [comment]
We can close this one for now. Since with the preseg passes we are no longer segment across `cat` any more.

Though suboptimal segmentation still remains an issue for rope, where @naoyam 's scheduler change should be able to resolve that.
